flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/StreamOperatorStateHandlerTest.java

```diff
@@ -76,12 +79,16 @@ public void testFailingBackendSnapshotMethod() throws Exception {
 			RunnableFuture<SnapshotResult<KeyedStateHandle>> keyedStateRawFuture = new CancelableFuture<>();
 			RunnableFuture<SnapshotResult<OperatorStateHandle>> operatorStateManagedFuture = new CancelableFuture<>();
 			RunnableFuture<SnapshotResult<OperatorStateHandle>> operatorStateRawFuture = new CancelableFuture<>();
+			RunnableFuture<SnapshotResult<StateObjectCollection<InputChannelStateHandle>>> inputChannelStateFuture = new CancelableFuture<>();
+			RunnableFuture<SnapshotResult<StateObjectCollection<ResultSubpartitionStateHandle>>> resultSubpartitionStateFuture = new CancelableFuture<>();
 
 			OperatorSnapshotFutures operatorSnapshotResult = new OperatorSnapshotFutures(
 				keyedStateManagedFuture,
 				keyedStateRawFuture,
 				operatorStateManagedFuture,
-				operatorStateRawFuture);
+				operatorStateRawFuture,
+				inputChannelStateFuture,
```

Comment:
Do we need to test that the `inputChannelStateFuture` and `resultSubpartitionStateFuture` is cancelled below?

Another question: do we have a test to cover that `inputChannelState` or `resultSubpartitionState` failed will fail the whole checkpoint?

Reply:
> Do we need to test that the inputChannelStateFuture and resultSubpartitionStateFuture is cancelled below?

Yes, that makes sense.

> do we have a test to cover that inputChannelState or resultSubpartitionState failed will fail the whole checkpoint?

AFAIK we don't have such tests. But there are tests that check that all the handles were passed to the CheckpointCoordinator, including the new handles. This means that TaskStateSnapshot won't be constructed if any of these futures fail.